### PR TITLE
stab setting to better Accurate.

### PR DESCRIPTION
**Pull request by yutayu**
I forgot this script. Sorry.
I cannot write script.I ask to shape new script you.

### CODE REVIEW

**Comment by yutayu** [comment]
>stab setting to better Accurate
All stab setting is necessary for accurate calculation.
For example, atm overhead is 5bytes per 53bytes.
ethenet overhead is 14bytes per each packet.

**Comment by robinsmidsrod** [comment]
@yutayu I have no idea what you're trying to achieve with this PR. I understand that the _stab_ qdisc does something related to table sizing depending on your link layer, but I don't see what it actually improves. Please explain more in detail, in better English.

**Comment by yutayu** [comment]
Since tc normally only calculates the size of an IP packet, the upper limit of the bandwidth must be set as low as 20% to allow for overhead, otherwise the actual link speed will be exceeded.
But there is a way to calculate the overhead exactly.
Since the stab setting is calculated in consideration of the overhead in advance, the speed of the link may be set as the upper limit of the bandwidth.

**Comment by yutayu** [comment]
Did you understand in my English?　ｌol
A bandwidth limit without a stub configuration cannot use all the bandwidth.
Stub configuration avoids wasted bandwidth.

**Comment by yutayu** [comment]
If you set the linklayer in the stab setting to atm, a packet of 1500 bytes will be calculated at 1696 bytes plus the overhead and will use all of the configured bandwidth.
If you leave it at 1500 bytes, you will need about 20% more than the configured bandwidth.

**Comment by yutayu** [comment]
I will try to persuade you for a better script.
What's New
This script also allows you to configure the stab feature so that you don't have to worry about overhead when setting the bandwidth limit.
For example, you can set 8 Mbit to use the 10 Mbit band. Otherwise the modem will cause buffer bloat.
Using the stab function, you only need to set about 9.5 Mbit, and you can calculate the queue correctly.

**Comment by yutayu** [comment]
Using the stab configuration, the ack packet size is calculated to be 106 bytes in qdisc instead of 40 bytes.
40 + overhead = (48 bytes + 5 bytes overhead) * 2
Adjustment is required.

**Comment by robinsmidsrod** [comment]
@yutayu If you intend to have all of the variables in the script, then you should use them to calculate the correct values in the `$stab` variable based on the type of link layer, according to the recommendations in the tc-stab man page, which you can read here: http://linuxcommandlibrary.com/man/tc-stab.html

You should use Bash math to actually calculate the the values, and I think the typical PPPoE ADSL/VDSL over ATM (which I think is one of the most common ways to connect using copper) connection should be the default in the script.

Also, can you squash your git commits into a single one and add some descriptive comments (with proper English grammar) that describes briefly how to calculate the stab values and points to the tc-stab man page for more details?

**Comment by yutayu** [comment]
I will try except bash calc.

**Comment by yutayu** [comment]
Please see newest.

File: supershaper.init [hunk 1]
```diff
@@ -108,6 +108,14 @@ TC=tc
 # Where to store flow labels, accessible for other tools
 FLOWS_FILENAME="/run/tc-flows.csv"
 
+# Stab setting
+# Configure a stab to calculate a more accurate queue size and avoid wasting bandwidth.
+# PPPoE ADSL/VDSL over ATM
+STAB="stab linklayer ethernet overhead 40"
```
Comment (robinsmidsrod) [inline supershaper.init]: @yutayu This number `40` is what I call a "magic number". You should include (in a comment) how it was calculated. Same for the number `10` below.

File: supershaper.init [hunk 2]
```diff
@@ -118,7 +126,9 @@ UPLINK=$((UPLINK_PERCENT*UPLINK_BW/100))
 # Which qdisc to use for HFSC leaves
 # Turns out that the old sfq seems to work better than fq_codel in
 # conjunction with HFSC Try out both and decide for yourself
-LEAF_QDISC="sfq divisor 65536 headdrop"
+# Quantum must be greater than or equal to MTU. Also, on ATM lines, Quantum should be a multiple of (48 + 5) = 53.
+# If MTU = 1500 , 1500 + 40 < 48 * 33 = 1584 , 33 * ( 48 + 5 ) = 1749.
```
Comment (robinsmidsrod) [inline supershaper.init]: @yutayu If you intend the SFQ quantum to be changed by the user, it should be a variable above line 119.

That calculation in the comment contains too many magic numbers. Please explain what those numbers are, like IP header, ATM header, PPP header, etc.

**Comment by yutayu** [comment]
How?

**Comment by robinsmidsrod** [comment]
> How?

How what? What was this a reference to? If you're referring to "squash commits", then I'm talking about using git rebase to change all of the different commits on this branch into a single commit. Also rewrite the commit message into something that is consistent with earlier commit messages.

**Pull request closed without merging**